### PR TITLE
`runBrokenLinks`

### PR DESCRIPTION
**Pull request by adamfweidman**
There are a few refactoring changes that happened in this PR but the function to review is `runBrokenLinks` and assosiated helpers in `broken_links.ts` 

There are 3 new helper functions in different files: `checkLinks`, `openNewPage`, and `closeBrowser`, `shuffleAndTruncate` have also been written.

### CODE REVIEW

File: packages/synthetics-sdk-broken-links/src/link_utils.ts [hunk 1]
```diff
@@ -302,6 +307,8 @@ export function setDefaultOptions(
     // per_link_options and linkorder are handled below
     if (
       !(optionName in inputOptions) ||
+      // eslint-disable-next-line  @typescript-eslint/no-explicit-any
+      (inputOptions as any)[optionName] === undefined ||
```
Comment (dkoss) [inline packages/synthetics-sdk-broken-links/src/link_utils.ts:311]: wait, how does this work?

File: packages/synthetics-sdk-broken-links/src/link_utils.ts [hunk 2]
```diff
@@ -520,5 +527,69 @@ export function createSyntheticResult(
     end_time: new Date().toISOString(),
   };
 
+  console.log('syn_result');
```
Comment (dkoss) [inline packages/synthetics-sdk-broken-links/src/link_utils.ts:530]: remove

File: packages/synthetics-sdk-broken-links/src/link_utils.ts [hunk 3]
```diff
@@ -520,5 +527,69 @@ export function createSyntheticResult(
...
+  link_limit: number,
+  link_order: BrokenLinksResultV1_BrokenLinkCheckerOptions_LinkOrder
+): LinkIntermediate[] {
+  // shuffle links if link_order is `RANDOM` and truncate to link_limit
+  // const links_to_follow: LinkIntermediate[] = [];
+
+  // Shuffle the links if link_order is RANDOM, or copy the original array
+  const linksToFollow =
+    link_order === BrokenLinksResultV1_BrokenLinkCheckerOptions_LinkOrder.RANDOM
+      ? [...links].sort(() => Math.random() - 0.5)
```
Comment (dkoss) [inline packages/synthetics-sdk-broken-links/src/link_utils.ts:590]: very good!

File: packages/synthetics-sdk-broken-links/src/link_utils.ts [hunk 4]
```diff
@@ -520,5 +527,69 @@ export function createSyntheticResult(
...
+ * @param link_order - Whether or not to shuffle links (enum value).
+ * @returns A new array of links that have been truncated based on the `link_limit`.
+ */
+export function shuffleAndTruncate(
+  links: LinkIntermediate[],
+  link_limit: number,
+  link_order: BrokenLinksResultV1_BrokenLinkCheckerOptions_LinkOrder
+): LinkIntermediate[] {
+  // shuffle links if link_order is `RANDOM` and truncate to link_limit
+  // const links_to_follow: LinkIntermediate[] = [];
```
Comment (dkoss) [inline packages/synthetics-sdk-broken-links/src/link_utils.ts:585]: remove the 2nd line of this comment

File: packages/synthetics-sdk-broken-links/test/unit/link_utils.spec.ts [hunk 5]
```diff
@@ -549,4 +553,40 @@ describe('GCM Synthetics Broken Links Utilies', async () => {
...
+      { target_url: 'link2', html_element: '', anchor_text: '' },
+      { target_url: 'link3', html_element: '', anchor_text: '' },
+      { target_url: 'link4', html_element: '', anchor_text: '' },
+      { target_url: 'link5', html_element: '', anchor_text: '' },
+    ];
+    const random =
+      BrokenLinksResultV1_BrokenLinkCheckerOptions_LinkOrder.RANDOM;
+    const firstN =
+      BrokenLinksResultV1_BrokenLinkCheckerOptions_LinkOrder.FIRST_N;
+    it('shuffles links when link_order is RANDOM', () => {
```
Comment (dkoss) [inline packages/synthetics-sdk-broken-links/test/unit/link_utils.spec.ts:569]: a little bit of padding goes a long way, add a new line before each `it` in this file?

File: packages/synthetics-sdk-broken-links/test/unit/broken_links_main.spec.ts [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+  runBrokenLinks,
+  BrokenLinkCheckerOptions,
+} from '../../src/broken_links_main';
+const path = require('path');
+
+describe('runBrokenLinks', async () => {
+  const status_class_2xx: ResponseStatusCode = {
+    status_class: ResponseStatusCode_StatusClass.STATUS_CLASS_2XX,
+  };
+  it('exits early if origin_url is false', async () => {
```
Comment (dkoss) [inline packages/synthetics-sdk-broken-links/test/unit/broken_links_main.spec.ts:35]: rename test description, `exits early if origin_url is in an incorrect format`?

File: packages/synthetics-sdk-broken-links/test/unit/broken_links_main.spec.ts [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+    )}`;
+    // Create a stub for puppeteer.launch to simulate launching a browser
+    const inputOptions = {
+      origin_url: origin_url,
+      query_selector_all: 'a[src], img',
+      get_attributes: ['href', 'src'],
+    };
+
+    const result = await runBrokenLinks(inputOptions);
+    console.log(result);
```
Comment (dkoss) [inline packages/synthetics-sdk-broken-links/test/unit/broken_links_main.spec.ts:77]: remove console log

File: packages/synthetics-sdk-broken-links/test/unit/broken_links_main.spec.ts [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+  ResponseStatusCode,
+  ResponseStatusCode_StatusClass,
+} from '@google-cloud/synthetics-sdk-api';
+import {
+  runBrokenLinks,
+  BrokenLinkCheckerOptions,
+} from '../../src/broken_links_main';
+const path = require('path');
+
+describe('runBrokenLinks', async () => {
```
Comment (dkoss) [inline packages/synthetics-sdk-broken-links/test/unit/broken_links_main.spec.ts:31]: I would test that branching logic that is specifically found only within runBrokenLinks is captured in these tests. Understandably, we cant test all the logic in the functions that runBrokenLinks calls, but there's behavior occurring between each step that doesnt seem to be completely captured across the board.

File: packages/synthetics-sdk-broken-links/src/broken_links_func.ts [hunk 9]
```diff
@@ -172,6 +113,37 @@ export async function retrieveLinksFromPage(
...
+  page: Page,
+  links: LinkIntermediate[],
+  options: BrokenLinksResultV1_BrokenLinkCheckerOptions
+): Promise<BrokenLinksResultV1_SyntheticLinkResult[]> {
+  const followed_links: BrokenLinksResultV1_SyntheticLinkResult[] = [];
+  for (const link of links) {
+    try {
+      followed_links.push(await checkLink(page, link, options));
+    } catch (err) {
+      if (err instanceof Error) process.stderr.write(err.message);
```
[the remark this answers is not in the document]
Comment (adamfweidman) [reply packages/synthetics-sdk-broken-links/src/broken_links_func.ts:105]: Any error thrown by the actual link navigation is caught by in `fetchLink`, this extra try-catch is in order to surround all asynchronous calls with a try-catch in case something totally unexpected happens.

Meaning if an error was ever thrown by checkLink that isn't from the link navigation we don't know how to handle it.

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -8,7 +8,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        node: [14, 16, 18, 20]
+        node: [18, 20]
```
[the remark this answers is not in the document]
Comment (adamfweidman) [reply .github/workflows/ci.yml:11]: https://github.com/GoogleCloudPlatform/synthetics-sdk-nodejs/pull/43

File: packages/synthetics-sdk-broken-links/src/broken_links_main.ts [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+      options.link_limit!,
+      options.link_order
+    );
+
+    // create new page to be used for all scraped links
+    const page = await openNewPage(browser);
+    // check all links
+    followed_links.push(...(await checkLinks(page, linksToFollow, options)));
+
+    await closeBrowser(browser);
```
[the remark this answers is not in the document]
Comment (adamfweidman) [reply packages/synthetics-sdk-broken-links/src/broken_links_main.ts:127]: done.

File: packages/synthetics-sdk-broken-links/src/link_utils.ts (hunk 3 above)
Comment (adamfweidman) [reply packages/synthetics-sdk-broken-links/src/link_utils.ts:590]: :)

File: packages/synthetics-sdk-broken-links/src/link_utils.ts [hunk 12]
```diff
@@ -126,10 +126,15 @@ export function checkStatusPassing(
...
+export function validateInputOptions(
+  inputOptions: BrokenLinkCheckerOptions
+): BrokenLinkCheckerOptions {
   if (!inputOptions.origin_url) {
     throw new Error('Missing origin_url in options');
-  } else if (!inputOptions.origin_url.startsWith('http')) {
+  } else if (
+    !inputOptions.origin_url.startsWith('http') &&
+    !inputOptions.origin_url.endsWith('.html')
+  ) {
```
[the remark this answers is not in the document]
Comment (adamfweidman) [reply packages/synthetics-sdk-broken-links/src/link_utils.ts:137]: fixed

File: packages/synthetics-sdk-broken-links/src/handlers.ts [hunk 13]
```diff
@@ -12,7 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-import { runBrokenLinks, BrokenLinkCheckerOptions } from './broken_links';
```
[the remark this answers is not in the document]
Comment (adamfweidman) [reply packages/synthetics-sdk-broken-links/src/handlers.ts:15]: renamed `broken_links_main --> broken_links` and `broken_links_func --> navigation_functionality`

File: packages/synthetics-sdk-broken-links/src/link_utils.ts (hunk 1 above)
Comment (adamfweidman) [reply packages/synthetics-sdk-broken-links/src/link_utils.ts:312]: this extra check is because 
{ origin_url: 'fake-link } is different than { origin_url: 'fake-link, link_limit: undefined }

`optionName in inputOptions` would evaluate to true in the 2nd case so I needed to add this statement

File: packages/synthetics-sdk-broken-links/src/broken_links_main.ts [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+  STATUS_CLASS_UNSPECIFIED = 'STATUS_CLASS_UNSPECIFIED',
+  STATUS_CLASS_1XX = 'STATUS_CLASS_1XX',
+  STATUS_CLASS_2XX = 'STATUS_CLASS_2XX',
+  STATUS_CLASS_3XX = 'STATUS_CLASS_3XX',
+  STATUS_CLASS_4XX = 'STATUS_CLASS_4XX',
+  STATUS_CLASS_5XX = 'STATUS_CLASS_5XX',
+  STATUS_CLASS_ANY = 'STATUS_CLASS_ANY',
+}
+
+export async function runBrokenLinks(
```
[the remark this answers is not in the document]
Comment (adamfweidman) [reply packages/synthetics-sdk-broken-links/src/broken_links_main.ts]: It is not quite as simple as that but a lot of functionality has been broken down into separate helpers. So that the try block is more readable. 

I think now it is still clear what is happening without being as complex:

```
process inputs
checkOriginLink
    return if it failed
scrape links
check scraped links
return
```

File: packages/synthetics-sdk-broken-links/test/unit/broken_links_main.spec.ts (hunk 8 above)
Comment (adamfweidman) [reply packages/synthetics-sdk-broken-links/test/unit/broken_links_main.spec.ts]: I ran the code coverage. the only unchecked branch was the wait_for_selector but I added a test for that.

File: packages/synthetics-sdk-broken-links/src/broken_links_func.ts [hunk 15]
```diff
@@ -172,6 +113,37 @@ export async function retrieveLinksFromPage(
...
+ */
+export async function checkLinks(
+  page: Page,
+  links: LinkIntermediate[],
+  options: BrokenLinksResultV1_BrokenLinkCheckerOptions
+): Promise<BrokenLinksResultV1_SyntheticLinkResult[]> {
+  const followed_links: BrokenLinksResultV1_SyntheticLinkResult[] = [];
+  for (const link of links) {
+    try {
+      followed_links.push(await checkLink(page, link, options));
```
[the remark this answers is not in the document]
Comment (adamfweidman) [reply packages/synthetics-sdk-broken-links/src/broken_links_func.ts]: I can't explain why I am getting the error that I am getting right now but previously it worked locally.

By using the pagePool we are able to run the requests in parallel without signfiiantly delaying any single link request (which can be a concern given the individual link_timeouts).

The time it takes to open a new page is not negligible, hence the creation of pagePool. For 60 links I found that 5 pages was faster than 10. I felt that therefore a good way to limit the concurrency is to limit the number of concurrent pages is link_limit / 10.

I spent the majority of Friday trying to figure out the `TypeError: Unknown file extension ".ts" ` issue but I don't know how.

File: .github/workflows/ci.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (dkoss) [reply .github/workflows/ci.yml:11]: TY!

File: packages/synthetics-sdk-broken-links/package.json [hunk 16]
```diff
@@ -40,6 +40,7 @@
   },
   "dependencies": {
     "@google-cloud/synthetics-sdk-api": "google-cloud-synthetics-sdk-api-0.1.0.tgz",
+    "plimit-lit": "^1.5.0",
```
Comment (dkoss) [inline packages/synthetics-sdk-broken-links/package.json:43]: why is this added?

File: packages/synthetics-sdk-broken-links/package.json [hunk 17]
```diff
@@ -40,6 +40,7 @@
   },
   "dependencies": {
     "@google-cloud/synthetics-sdk-api": "google-cloud-synthetics-sdk-api-0.1.0.tgz",
```
Comment (dkoss) [inline packages/synthetics-sdk-broken-links/package.json:42]: side node - can this be updated to target an actual version of this dependency now?

Comment (adamfweidman) [reply packages/synthetics-sdk-broken-links/package.json:42]: All broken link synthetic api information is in broken_links_final. It has not been pushed to main / npm so the dependency will stay as a tar file until we merge into main.

File: packages/synthetics-sdk-broken-links/package.json (hunk 16 above)
Comment (adamfweidman) [reply packages/synthetics-sdk-broken-links/package.json:43]: This is what allows to limit concurrency (see `checkLinks`)

My comment from above that probably got lost in the ether:

https://github.com/GoogleCloudPlatform/synthetics-sdk-nodejs/pull/41#discussion_r1320365682

File: packages/synthetics-sdk-broken-links/package.json (hunk 17 above)
Comment (dkoss) [reply packages/synthetics-sdk-broken-links/package.json:42]: got it, thanks!

File: packages/synthetics-sdk-broken-links/src/broken_links.ts [hunk 18]
```diff
@@ -56,59 +68,124 @@ export enum StatusClass {
...
+ * Scrapes links from the origin page and returns them.
+ * If applicable:
+ *     - wait for `options.wait_for_selector` element before scraping.
+ *     - shuffle and truncate based on `options`
+ *
+ * @param originPage - The Puppeteer page object representing the origin page.
+ * @param options - The broken link checker options.
+ * @returns An array of scraped links in accordance with link_limit and link_order.
+ */
+async function scrapeLinks(
```
Comment (dkoss) [inline packages/synthetics-sdk-broken-links/src/broken_links.ts:156]: should this be in link utils? (by nature of being called, scrape links :))

File: packages/synthetics-sdk-broken-links/src/navigation_func.ts [hunk 19]
```diff
@@ -77,6 +80,54 @@ export async function retrieveLinksFromPage(
...
+  links: LinkIntermediate[],
+  options: BrokenLinksResultV1_BrokenLinkCheckerOptions
+): Promise<BrokenLinksResultV1_SyntheticLinkResult[]> {
+  const followed_links: BrokenLinksResultV1_SyntheticLinkResult[] = [];
+  const limit = pLimit(options.link_limit! / 10);
+
+  // Initialize a pool of pages
+  const pagePool: Page[] = [];
+
+  await Promise.all(
```
Comment (dkoss) [inline packages/synthetics-sdk-broken-links/src/navigation_func.ts:106]: I think that we talked about this in another thread - im still confused as to what happens if any of these links end up failing? Doesnt this exit mid-check without going through all the links?

I'm thinking through scenarios where for some reason, 99% of links work except 1 which crashes puppeteer for some reason. Do we really want to abandon ship and not return that information? LMK if im missing something

**Review by dkoss — APPROVED** [review]
Left one comment - let's make sure to close on all of the sort-of not happy paths during bug bash sessions. We need to make sure that the generic error is a last-resort type situation, and to exit as gracefully as possible even when the unexpected happens.

File: packages/synthetics-sdk-broken-links/src/navigation_func.ts (hunk 19 above)
Comment (adamfweidman) [reply packages/synthetics-sdk-broken-links/src/navigation_func.ts:106]: Yes and no...

- The call to page.goto(), which is the actual navigation is done in a function called `fetchLink()`, this has its own try-catch which handles if an error is thrown.

- checkLink() has no other asynchronous call, the only "risky" call is to page.goto(). `checkLink()` should never throw an unexpected error and this catch block should never be executed, this is just incase. Also I don't think  there is a  way to handle this error other than throwing or ignoring and logging (which is not severe enough for the scenario in my opinion).

File: packages/synthetics-sdk-broken-links/src/broken_links.ts (hunk 18 above)
Comment (adamfweidman) [reply packages/synthetics-sdk-broken-links/src/broken_links.ts:156]: I was trying to keep all the helpers that are there purely to make `runBrokenLinks` more readable in broken_links.ts

What do you think? These methods are not independently tested but tested through the `runBrokenLinks` tests

File: packages/synthetics-sdk-broken-links/src/link_utils.ts [hunk 20]
```diff
@@ -254,3 +258,102 @@ export function createSyntheticResult(
 
   return synthetic_result;
 }
+
+/**
+ * Opens a new Puppeteer page within the provided browser instance, disables caching, and returns the created page.
```
Comment (dkoss) [inline packages/synthetics-sdk-broken-links/src/link_utils.ts:263]: move browser based functions to navigation_func.ts

File: packages/synthetics-sdk-broken-links/src/navigation_func.ts [hunk 21]
```diff
@@ -77,6 +79,53 @@ export async function retrieveLinksFromPage(
...
+    links.map(async (link) => {
+      await limit(async () => {
```
Comment (dkoss) [inline packages/synthetics-sdk-broken-links/src/navigation_func.ts:107]: move concurrency to a separate PR

File: packages/synthetics-sdk-broken-links/test/unit/broken_links.spec.ts [hunk 22]
```diff
@@ -0,0 +1,187 @@
...
+    const genericResult = syntheticResult.synthetic_generic_result_v1;
+    expect(genericResult).to.be.exist;
+    expect(genericResult?.ok).to.be.false;
+    expect(genericResult?.generic_error?.error_type).to.equal('Error');
+    expect(genericResult?.generic_error?.error_message).to.equal(
+      'origin_url must be a string that starts with `http`'
+    );
+  }).timeout(10000);
+
+  it('returns generic_result when waitForSelector exceeds deadline', async () => {
```
Comment (dkoss) [inline packages/synthetics-sdk-broken-links/test/unit/broken_links.spec.ts:64]: Change from generic error to one that says that the origin url passed === false, with info explaining that the content that we're looking for is not present.

File: packages/synthetics-sdk-broken-links/src/broken_links_func.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (adamfweidman) [reply packages/synthetics-sdk-broken-links/src/broken_links_func.ts]: to be discussed in later PR

**Review by dkoss — APPROVED** [review]
